unison-src/transcripts-using-base/base.u

```diff
@@ -224,3 +234,133 @@ MVar.take = compose reraise take.impl
 MVar.read = compose reraise read.impl
 MVar.put = compose2 reraise put.impl
 MVar.swap = compose2 reraise MVar.swap.impl
+
+-- Some functions for using a sorted list as a map
+bSearch : [(a,b)] -> a -> Optional b
+bSearch m k =
+  find l u =
+    i = (l+u)/2
+    if l >= u then None
+    else match at i m with
+      Some (k', v)
+        | k == k' -> Some v
+        | k > k' -> find (i+1) u
+        | k < k' -> find l i
+      None -> None
+  find 0 (size m)
+
+bContains : [(a,b)] -> a -> Boolean
+bContains m k = match bSearch m k with
+  Some _ -> true
+  None -> false
+
+bSplit : [(a,b)] -> a -> ([(a,b)], [(a,b)])
+bSplit m k =
+  find l u =
+    i = (l+u)/2
+    if l >= u then (List.take l m, List.drop l m)
+    else match at i m with
+      Some (k', _)
+        | k == k' -> (List.take i m, List.drop (i+1) m)
+        | k > k' -> find (i+1) u
+        | k < k' -> find l i
+      None -> (m, [])
+  find 0 (size m)
+
+bInsert : [(a,b)] -> a -> b -> [(a,b)]
+bInsert m k v = match bSplit m k with
+  (pre, post) -> pre ++ [(k,v)] ++ post
+
+bSort : [(a,b)] -> [(a,b)]
+bSort = foldLeft (acc -> uncurry (bInsert acc)) []
+
+-- Finding transitive dependencies
+crawl found = cases
+  [] -> found
+  ln +: rest
+    | Text.take 2 (Link.Term.toText ln) == "##" -> crawl found rest
+    | bContains found ln -> crawl found rest
```

Comment:
bContains requires a sorted list right?

Is `found` guaranteed to be sorted?
Also would likely be worth mentioning the sorting invariants that are required on each of the functions, e.g. `bContains`, `bSearch`

Reply:
Okay, I added some comments to these.